feat: augments

phacUFPE opened the pull request:
# Description

Implementing Augments system.
Please check the system in the following links:
https://tibia.fandom.com/wiki/Augments
https://www.tibia.com/news/?subtopic=newsarchive&id=7356


What must be done:
- [x] Add augments to all items that have it;
- [x] Parse augments from items.xml;
- [x] Add default value from config lua for augments without value;
- [x] Store augments into item;
- [x] Show augments on market;
- [x] Display augments on item look;
- [x] Apply augments cooldown on spell usage;
- [x] Apply augments that increase damage on spell usage;
- [x] Apply augments life/mana leech on spell usage;
- [x] Test against players (I ask for the community help to test);
- [x] Test against monsters (I ask for the community help to test).


Suggestions are welcome.

![image](https://github.com/opentibiabr/canary/assets/32341050/fb15052c-974f-44c0-ae05-5ba223cf96da)

![image](https://github.com/opentibiabr/canary/assets/32341050/b490c080-a312-46a8-88bc-ef06b7256475)

![image](https://github.com/opentibiabr/canary/assets/32341050/b4b46456-bb98-4bd4-8ad1-a613fcb09304)

# How to configure an item augment:
## Supported augment types are:
* "critical extra damage",
* "life leech", 
* "mana leech",
* "cooldown",
* "increased damage", (don't need to set value, but can be customized for a specific item)
* "strong impact", (don't need to set value, but can be customized for a specific item)
* "powerful impact" (don't need to set value, but can be customized for a specific item)

## Example of augment that have default value::
```xml
<attribute key="augments" value="1">
	<attribute key="spell name" value="augment type" />
</attribute>
```
#### NOTE: The key value of the key "augments" is to enable (1) and disable (0) the augments of the item.

## Example of augment that don't have default value, the value must be set:
#### NOTE: The values represent percentagel, except for cooldown, it should be fill in miliseconds (120 seconds = 120000).
```xml
<attribute key="augments" value="1">
   <attribute key="energy wave" value="life leech">
     <attribute key="value" value="5" />
   </attribute>
</attribute>
```

## How to customize an augment that already has a default value?
Just add attribute tag with the key value and the value that you which, example:

```xml
<attribute key="augments" value="1">
   <attribute key="energy wave" value="powerful impact">
     <attribute key="value" value="15" />
   </attribute>
</attribute>
```

phacUFPE committed:
feat: augments
changes:
- Added AugmentTypeNames map in tools
- Added AugmentTypes_t to items_definitions
- Added augments properties on items
- Added Augments type using std::map<std::string, std::map<AugmentTypes_t, uint8_t>>

luanluciano93 commented:
https://github.com/opentibiabr/canary/blob/main/src/items/functions/item/item_parse.cpp#L80
```C++
ItemParse::parseAugment(tmpStrValue, attributeNode, valueAttribute, itemType);
```

in parse https://github.com/opentibiabr/canary/blob/9f74b27eff36c4e2774e635dc52baf8655e47c6b/src/items/functions/item/item_parse.hpp#L249
```c++
const phmap::flat_hash_map<std::string, AugmentsTypes_t> AugmentsTypeMap = {
	{ "powerful impact", AUGMENT_VALUE_POWERFUL_IMPACT },
	{ "strong impact", AUGMENT_VALUE_STRONG_IMPACT },
	{ "increased damage", AUGMENT_VALUE_INCREASED_DAMAGE },
	{ "cooldown", AUGMENT_VALUE_COOLDOWN },
	{ "critical extra damage", AUGMENT_VALUE_CRITICALHITDAMAGE },
	{ "life leech", AUGMENT_VALUE_LIFELEECHAMOUNT },
	{ "mana leech", AUGMENT_VALUE_MANALEECHAMOUNT },
};
```

luanluciano93 commented on src/items/functions/item/item_parse.cpp:
```diff
@@ -870,6 +870,47 @@ void ItemParse::parseImbuement(const std::string &tmpStrValue, pugi::xml_node at
+	}
+
+	// Check if the augments value is 1 or 0 (1 = enable - 0 = disable)
+	if (pugi::cast<bool>(valueAttribute.value())) {
```
I think this is necessary

luanluciano93 commented on src/items/functions/item/item_parse.cpp:
```diff
@@ -870,6 +870,47 @@ void ItemParse::parseImbuement(const std::string &tmpStrValue, pugi::xml_node at
+				}
+			} else {
+				g_logger().warn("[ParseImbuement::initParseImbuement] - Unknown type: {}", valueAttribute.as_string());
+			}
```
```suggestion
			auto itemMap = AugmentsTypeMap.find(asLowerCaseString(subKeyAttribute.as_string()));
			if (itemMap != AugmentsTypeMap.end()) {
				AugmentTypes_t augmentType = getImbuementType(asLowerCaseString(subKeyAttribute.as_string()));
				if (augmentType != IMBUEMENT_NONE) {
					itemType.setImbuementType(augmentType, pugi::cast<uint16_t>(subValueAttribute.value()));
					continue;
				}
			} else {
				g_logger().warn("[ParseAugment::initParseImbuement] - Unknown type: {}", valueAttribute.as_string());
			}
```

phacUFPE replied:
I`m current adding this part, thanks!

phacUFPE committed:
feat: augments
changes:
- Added ITEM_PARSE_AUGMENT into item parse mapping
- Update parseAugment item parse (working)

github-actions[bot] committed:
Code format - (Clang-format)

phacUFPE commented:
Update!

Item parse working and storing into item.

![image](https://github.com/opentibiabr/canary/assets/32341050/5f498d1e-392c-4a17-938a-2718a23991a3)

phacUFPE committed:
feat: augments
changes:
- Improvements
- Parse augment description

phacUFPE committed:
Merge branch 'feat_weapons_augment' of https://github.com/phacUFPE/canary into feat_weapons_augment

github-actions[bot] committed:
Code format - (Clang-format)

phacUFPE commented:
Update!

Added item description.

![image](https://github.com/opentibiabr/canary/assets/32341050/43fedefe-ef88-4a73-af3f-8870b9def6df)

phacUFPE committed:
feat: augments
changes:
- Added augments to all items that have it

phacUFPE commented:
Update!

Added augments to all **sanguine** and **grand sanguine** items.

phacUFPE committed:
feat: augments
changes:
- Added config key to load default values for: increase damage, powerful impact and strong impact
- Added to item parse the load of the augment value for these augments that do not display description
- Added keys to config.lua.dist

github-actions[bot] committed:
Code format - (Clang-format)

phacUFPE committed:
feat: augments
changes:
- Added augment description parser on utils tools
- Added augment description in market

phacUFPE committed:
Merge branch 'feat_weapons_augment' of https://github.com/phacUFPE/canary into feat_weapons_augment

github-actions[bot] committed:
Code format - (Clang-format)

phacUFPE commented:
Update!

Added augments descriptions on market.

![image](https://github.com/opentibiabr/canary/assets/32341050/802eef13-c6a1-4178-af30-9c510a895b02)

phacUFPE committed:
feat: augments
changes:
- Added cooldown as miliseconds

phacUFPE committed:
feat: augments
changes:
- Added spell cooldown reduction to spells
- Added method to retrieve the augment items by spell name and/or by type
- Added spell method to get the spell cooldown reduction from the augment items

phacUFPE committed:
Merge branch 'feat_weapons_augment' of https://github.com/phacUFPE/canary into feat_weapons_augment

phacUFPE commented:
Update!

Cooldown augments working (except for avatar, avatar spells cooldown are calculated in the casting, after this PR I'll be investigating to fix it).

https://github.com/opentibiabr/canary/assets/32341050/cfaadfca-eea1-49d0-baed-35a21092c06d

tomasz1337 commented:
Nice one! i will do some testing today !!! Thanks a lot for that contribution

jbracovich commented:
have this warning is normal ?
![image](https://github.com/opentibiabr/canary/assets/165123587/9a221147-6c4b-4a10-a360-015fac97bd75)

phacUFPE commented:
> have this warning is normal ? ![image](https://private-user-images.githubusercontent.com/165123587/327962944-9a221147-6c4b-4a10-a360-015fac97bd75.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ4NTczODcsIm5iZiI6MTcxNDg1NzA4NywicGF0aCI6Ii8xNjUxMjM1ODcvMzI3OTYyOTQ0LTlhMjIxMTQ3LTZjNGItNGExMC1hMzYwLTAxNWZhYzk3YmQ3NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwNFQyMTExMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03N2IxZmMwNTNiMDQ3ZWZmODNmNTI3ZmU0Yzg2OTcyZmE1YWVlMTkyYjc2ZjM0OTM0ZTAzZGU1Nzc3YWZiYjAzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.HwVw95OVWozIdDBtK0i2Xl6j-1WEycKmNRocYRWoaVw)

No, did you updated your code with this PR ?

phacUFPE committed:
Merge branch 'main' into feat_weapons_augment

phacUFPE committed:
feat: augments
changes:
- Minor improvements

phacUFPE committed:
feat: augments
changes:
- Improvements
- Added damage augments to work on spell usage
- Added life leech and mana leech augments to work on spell usage

phacUFPE commented:
Update!

Augments working: Spell Cooldown (except for avatars, I'll be investigating and fixing in another PR), Spell Damages (Increased Damage, Strong Impact and Powerful Impact), Spell Life Leech, Mana Leech and Spell Critical Extra Damage (this was not tested).

Test with Spell Damage Augments between swords that don't have it and with one sword that have. (Spell damage limited to be easier to identify damage modifications).

https://github.com/opentibiabr/canary/assets/32341050/9294d8d4-52b0-40be-ab84-a2e89a1663ce


This PR needs more testing to check if nothing else in addition of augments was affected.

I'll be opening for review. Please report any bugs, suggestions to improve the time and memory efficiency or other.

phacUFPE committed:
feat: augments
changes:
- Removed deleted file

Mirkaanks commented:
Good Job!

I started doing it, but stopped out of sheer laziness.

phacUFPE commented:
This PR needs #2607 to the avatar cooldown reduction augments works.

phacUFPE committed:
feat: augments
changes:
- Added augments description cyclopedia items and player item inspection

phacUFPE commented:
Update!

Showing augments description in item cyclopedia and item inspection.

![image](https://github.com/opentibiabr/canary/assets/32341050/a94c7484-0a35-4232-91d7-061204cb5525)

![image](https://github.com/opentibiabr/canary/assets/32341050/2ae4b772-5221-49ae-a67e-c0feaa7828d7)

jbracovich commented:
![image](https://github.com/opentibiabr/canary/assets/165123587/020ea6e9-8c54-41b1-9055-6c8b7ef85408)
Good job bro !

pedrohfk commented:
I'm testing, apparently everything is ok, congratulations on the work!!

elsongabriel committed:
Merge branch 'main' into feat_weapons_augment

elsongabriel commented on config.lua.dist:
```diff
@@ -80,6 +80,13 @@ freeDepotLimit = 2000
+-- To customize the percentage of a particular item, please add to the item "augments" section on items.xml
+augmentIncreasedDamageDefault = 5 * 100 -- 5%
+augmentPowerfulImpactDefault = 7 * 100 -- 7%
+augmentStrongImpactDefault = 10 * 100 -- 10%
```
please remove spaces before commenting in the config.lua, you can put info into NOTE.
I recommend you change the config types to int (example: 5 to be 5%) and you'll do the conversion in your scripts, to avoid user bugs ..

elsongabriel commented on src/config/configmanager.cpp:
```diff
@@ -345,6 +345,9 @@ bool ConfigManager::load() {
 	loadIntConfig(L, WHITE_SKULL_TIME, "whiteSkullTime", 15 * 60 * 1000);
+	loadIntConfig(L, AUGMENT_INCREASED_DAMAGE_DEFAULT, "augmentIncreasedDamageDefault", 5 * 100);
+	loadIntConfig(L, AUGMENT_POWERFUL_IMPACT_DEFAULT, "augmentPowerfulImpactDefault", 10 * 100);
+	loadIntConfig(L, AUGMENT_STRONG_IMPACT_DEFAULT, "augmentStrongImpactDefault", 7 * 100);
```
the same thing here. put the value with only 1 - 100.

phacUFPE replied:
👍 
I'll do the same for the not default to be as integer between 1 - 100 to represent the percentage.

elsongabriel commented on src/items/functions/item/item_parse.cpp:
```diff
@@ -870,6 +871,67 @@ void ItemParse::parseImbuement(const std::string &tmpStrValue, pugi::xml_node at
+			<attribute key="value" value="5" /> -- WHEN NESTING AND ADDING VALUE COULD BE CUSTOMIZED ?
+		</attribute>
+	</attribute>
+	*/
```
Have you decided what's the best way? if yes, please remove this comment code.
you can put in config.lua NOTE, the URL to this PR, and create a "Documentation" for more information about this feature.

phacUFPE replied:
Yes, thanks. I forgot to remove this comment. Could you help me telling me where it's the best place to create a documentation of how to configure augments ?

elsongabriel replied:
it's a piece of simple information, and your PR is already very complete, you can check in the config.lua.dist the "Vip System" config.

phacUFPE replied:
> it's a piece of simple information, and your PR is already very complete, you can check in the config.lua.dist the "Vip System" config.

Thanks, documetation added.

elsongabriel commented on src/creatures/combat/spells.cpp:
```diff
@@ -630,6 +630,38 @@ void Spell::setWheelOfDestinyBoost(WheelSpellBoost_t boost, WheelSpellGrade_t gr
+				damage.manaLeech += augment->type == AUGMENT_MANALEECHAMOUNT ? augment->value : 0;
+				damage.criticalDamage += augment->type == AUGMENT_CRITICALHITDAMAGE ? augment->value : 0;
+				if (augment->type == AUGMENT_INCREASED_DAMAGE || augment->type == AUGMENT_POWERFUL_IMPACT || augment->type == AUGMENT_STRONG_IMPACT) {
+					const float augmentPercent = augment->value / 10000.0;
```
I think you can adjust your calculation here.

phacUFPE replied:
I'll reduce to percentage and adjust the calculation with the percentage.

elsongabriel reviewed (commented):
Nice job, code organized. Congratz!

phacUFPE committed:
feat: augments
changes:
- Suggestions applied
- Improvements

phacUFPE commented:
> Nice job, code organized. Congratz!

Thanks, suggestions and comments fixed.

phacUFPE committed:
feat: augments
changes:
- Added documentation

dudantas commented on config.lua.dist:
```diff
@@ -80,6 +80,48 @@ freeDepotLimit = 2000
+-- * "cooldown",
+-- * "increased damage", (don't need to set value)
+-- * "strong impact", (don't need to set value)
+-- * "powerful impact" (don't need to set value)
```
Things that are not related to the config are better to extract in a separate "readme.md" or document on the wiki, as this will "dirty" the config.lua a lot for things that have no direct connection with the configuration file. It is not the function of config.lua to teach how to use a system.

phacUFPE replied:
Elson already warned me, I'll be removing and adding to the PR.

dudantas commented on src/creatures/combat/spells.cpp:
```diff
@@ -630,6 +630,40 @@ void Spell::setWheelOfDestinyBoost(WheelSpellBoost_t boost, WheelSpellGrade_t gr
+void Spell::getCombatDataAugment(std::shared_ptr<Player> player, CombatDamage &damage) {
+	if (!(damage.instantSpellName).empty()) {
+		std::vector<std::shared_ptr<Item>> equippedAugmentItems = player->getEquippedAugmentItems();
+		for (std::shared_ptr<Item> item : equippedAugmentItems) {
```
Pass as a constant reference to avoid unnecessary copying
```suggestion
		const auto &equippedAugmentItems = player->getEquippedAugmentItems();
		for (const auto &item : equippedAugmentItems) {
```

dudantas commented on src/creatures/combat/spells.cpp:
```diff
@@ -630,6 +630,40 @@ void Spell::setWheelOfDestinyBoost(WheelSpellBoost_t boost, WheelSpellGrade_t gr
+			Augments augments = item->getAugmentsBySpellName(damage.instantSpellName);
+			for (auto &augment : augments) {
+				if (augment->type == AUGMENT_INCREASED_DAMAGE || augment->type == AUGMENT_POWERFUL_IMPACT || augment->type == AUGMENT_STRONG_IMPACT) {
+					const float augmentPercent = augment->value / 100.0;
```
Divisions by 0 can cause undefined behavior (crash) it may be interesting to check that "augment->value" is not 0 before dividing

```suggestion
					const float augmentPercent = augment->value ? augment->value / 100.0 : 0;
```

dudantas commented on src/creatures/combat/spells.cpp:
```diff
@@ -630,6 +630,40 @@ void Spell::setWheelOfDestinyBoost(WheelSpellBoost_t boost, WheelSpellGrade_t gr
+
+int32_t Spell::calculateAugmentSpellCooldownReduction(std::shared_ptr<Player> player) const {
+	int32_t spellCooldown = 0;
+	std::vector<std::shared_ptr<Item>> equippedAugmentItems = player->getEquippedAugmentItemsByType(AUGMENT_COOLDOWN);
```
```suggestion
	const auto &equippedAugmentItems = player->getEquippedAugmentItemsByType(AUGMENT_COOLDOWN);
```

dudantas commented on src/creatures/combat/spells.cpp:
```diff
@@ -630,6 +630,40 @@ void Spell::setWheelOfDestinyBoost(WheelSpellBoost_t boost, WheelSpellGrade_t gr
+int32_t Spell::calculateAugmentSpellCooldownReduction(std::shared_ptr<Player> player) const {
+	int32_t spellCooldown = 0;
+	std::vector<std::shared_ptr<Item>> equippedAugmentItems = player->getEquippedAugmentItemsByType(AUGMENT_COOLDOWN);
+	for (std::shared_ptr<Item> item : equippedAugmentItems) {
```
```suggestion
	for (const auto &item : equippedAugmentItems) {
```

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -4064,6 +4064,35 @@ std::vector<std::shared_ptr<Item>> Player::getAllInventoryItems(bool ignoreEquip
+	std::vector<std::shared_ptr<Item>> equippedAugmentItemsByType;
+	std::vector<std::shared_ptr<Item>> equippedAugmentItems = getEquippedItems();
+
+	for (std::shared_ptr<Item> item : equippedAugmentItems) {
```
```suggestion
	for (const auto &item : equippedAugmentItems) {
```

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -4064,6 +4064,35 @@ std::vector<std::shared_ptr<Item>> Player::getAllInventoryItems(bool ignoreEquip
+	std::vector<std::shared_ptr<Item>> equippedAugmentItems;
+	std::vector<std::shared_ptr<Item>> equippedItems = getEquippedItems();
+
+	for (std::shared_ptr<Item> item : equippedItems) {
```
```suggestion
	for (const auto &item : equippedItems) {
```

dudantas commented on src/items/item.cpp:
```diff
@@ -1437,6 +1437,22 @@ Item::getDescriptions(const ItemType &it, std::shared_ptr<Item> item /*= nullptr
+		if (augments.size() > 0) {
+			ss.str("");
+			uint8_t count = 0;
+			for (auto augment : augments) {
```
```suggestion
			for (const auto &augment : augments) {
```

dudantas commented on src/items/item.cpp:
```diff
@@ -1789,6 +1805,21 @@ Item::getDescriptions(const ItemType &it, std::shared_ptr<Item> item /*= nullptr
+		if (it.augments.size() > 0) {
+			ss.str("");
+			uint8_t count = 0;
+			for (auto augment : it.augments) {
```
```suggestion
			for (const auto &augment : it.augments) {
```

dudantas commented on src/items/item.cpp:
```diff
@@ -1897,6 +1928,35 @@ Item::getDescriptions(const ItemType &it, std::shared_ptr<Item> item /*= nullptr
+		  << "Augments: (";
+
+		uint8_t count = 0;
+		for (auto &augment : augments) {
```
```suggestion
		for (const auto &augment : augments) {
```

dudantas commented on src/items/items.hpp:
```diff
@@ -435,6 +442,35 @@ class Items {
 	}
 
+	static const std::string getAugmentNameByType(AugmentTypes_t augmentType) {
+		phmap::flat_hash_map<AugmentTypes_t, std::string> map = {
```
Add it as static to prevent the map from being recreated with each function call

```suggestion
		static phmap::flat_hash_map<AugmentTypes_t, std::string> map = {
```

dudantas commented on src/items/items.hpp:
```diff
@@ -435,6 +442,35 @@ class Items {
 		return dummys;
 	}
 
+	static const std::string getAugmentNameByType(AugmentTypes_t augmentType) {
```
Analyze the possibility of the string being a reference to avoid unnecessary copying

dudantas commented on src/items/items.hpp:
```diff
@@ -435,6 +442,35 @@ class Items {
+	}
+
+	static const bool isAugmentWithoutValueDescription(AugmentTypes_t augmentType) {
+		std::vector<AugmentTypes_t> vector = {
```
As previously mentioned, avoid creating a new vector during each function call to optimize performance.
```suggestion
		static std::vector<AugmentTypes_t> vector = {
```

dudantas commented on src/items/items_definitions.hpp:
```diff
@@ -606,9 +617,21 @@ enum ItemParseAttributes_t {
+	int32_t value;
+};
+
+using Augments = std::vector<std::shared_ptr<AugmentInfo>>;
```
I have reservations about using aliases, as they obscure the object types, which can complicate code review. Although aliases can simplify maintenance, they might pose challenges in a large, open-source project like Canary. 
But, I'm not opposing it, it's something particular to each person, as we don't define a model to be followed in the project.

phacUFPE replied:
I don't like either, I saw some aliases and I thought that was a pattern to follow, reverting.

dudantas commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -5604,7 +5604,21 @@ void ProtocolGame::sendMarketDetail(uint16_t itemId, uint8_t tier) {
+		if (it.augments.size() > 0) {
+			std::ostringstream augmentsString;
+			uint8_t count = 0;
+			for (auto &augment : it.augments) {
```
```suggestion
			for (const auto &augment : it.augments) {
```

dudantas commented on src/utils/tools.cpp:
```diff
@@ -1558,6 +1577,22 @@ void capitalizeWords(std::string &source) {
 	}
 }
 
+void capitalizeWordsIgnoringString(std::string &source, std::string stringToIgnore) {
```
The variable "stringToIgnore" can be passed a const reference

dudantas reviewed (changes requested):
I have a few suggestions that could enhance performance and address other issues. Please remember to test these changes before implementing them, as my review is based solely on the code, and I haven’t applied them locally. Additional adjustments might be necessary to prevent compilation errors.

I will conduct the review in a way that locks the PR, because some modifications are optional, while others, like unnecessary copying that directly impacts performance, or dividing by zero without checking the divisor, could potentially cause more serious issues. Feel free to reach out to me for another review when you're ready.

phacUFPE committed:
feat: augments
changes:
- Applying suggestions and comments

phacUFPE committed:
feat: augments
changes:
- Applying suggestions and comments

github-actions[bot] committed:
Code format - (Clang-format)

phacUFPE committed:
Merge branch 'feat_weapons_augment' of https://github.com/phacUFPE/canary into feat_weapons_augment

phacUFPE commented:
> I have a few suggestions that could enhance performance and address other issues. Please remember to test these changes before implementing them, as my review is based solely on the code, and I haven’t applied them locally. Additional adjustments might be necessary to prevent compilation errors.
> 
> I will conduct the review in a way that locks the PR, because some modifications are optional, while others, like unnecessary copying that directly impacts performance, or dividing by zero without checking the divisor, could potentially cause more serious issues. Feel free to reach out to me for another review when you're ready.

All fixed and working as expected.

phacUFPE committed:
Merge branch 'main' into feat_weapons_augment

phacUFPE committed:
feat: augments
- Added doc to config.lua.dist as requested by elsongabriel

phacUFPE committed:
feat: augments
- Added const auto &augments to avoid unnecessary copies

phacUFPE committed:
feat: augments
changes:
- Improvements

phacUFPE committed:
feat: augments
changes:
- Improvements

github-actions[bot] committed:
Code format - (Clang-format)

phacUFPE committed:
Merge branch 'feat_weapons_augment' of https://github.com/phacUFPE/canary into feat_weapons_augment

github-actions[bot] committed:
Code format - (Clang-format)

phacUFPE committed:
fix: strong impact value on config.lua.dist
changes:
- Updated augmentStrongImpactPercent from 100 to 10

phacUFPE committed:
Merge branch 'main' into feat_weapons_augment

phacUFPE committed:
Merge branch 'main' into feat_weapons_augment

phacUFPE committed:
Merge branch 'main' into feat_weapons_augment

phacUFPE committed:
Merge branch 'feat_weapons_augment' of https://github.com/phacUFPE/canary into feat_weapons_augment

dudantas committed:
improve: use enum cast to find types

fix: extract augment name in own variable

fix: remove continue

phacUFPE committed:
fix: add spaces before capital letters
changes:
- Called method to add spaces before capital letters on item augment description
